### PR TITLE
rumqttc: Enable native tls `TlsConnector` support.

### PR DESCRIPTION
**Pull request by xiaocq2001**
## Type of change

Resolves https://github.com/bytebeamio/rumqtt/issues/866.

New feature (non-breaking change which adds functionality)
* `Transport::tls_with_config` now accepts `native_tls::TlsConnectorBuilder` instance as input, to customize options

## Checklist:

- [ ] Formatted with `cargo fmt`
- [ ] Make an entry to `CHANGELOG.md` if it's relevant to the users of the library. If it's not relevant mention why.

### CODE REVIEW

File: rumqttc/src/lib.rs [hunk 1]
```diff
@@ -315,7 +319,7 @@ impl Transport {
 }
 
 /// TLS configuration method
-#[derive(Clone, Debug)]
+#[derive(Clone)]
```
Comment (xiaocq2001) [inline rumqttc/src/lib.rs:322]: derive of Debug is removed because TlsConnectorBuilder has no Debug trait, not sure if there is better idea for it.

**Comment by xiaocq2001** [comment]
Considering that the `Native` could be used to simplify native configuration to use OS settings, and there may be compatible issue, I added new `NativeBuild` to accept builder as input.

**Review by de-sh — APPROVED** [review]
Hopefully no one depends on `TlsConfiguration` being `Debug`, `Transport` ain't so LGTM

File: rumqttc/examples/tls_native.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    color_backtrace::install();
+
+    let mut mqttoptions = MqttOptions::new("test-1", "test.mosquitto.org", 8883);
+    mqttoptions.set_keep_alive(std::time::Duration::from_secs(5));
+
+    // Use native-tls to load root certificates from the operating system.
+    let mut builder = native_tls::TlsConnector::builder();
+    let pem = vec![1, 2, 3];
+    // let pem = include_bytes!("native-tls-cert.pem");
+    let cert = native_tls::Certificate::from_pem(pem).unwrap();
```
Comment (de-sh) [inline rumqttc/examples/tls_native.rs:19]:
```suggestion
    let cert = native_tls::Certificate::from_pem(&pem).unwrap();
```

**Comment by de-sh** [comment]
Since we don't do much with the `Builder` under the hood, why not just ask the user to build it first?

**Comment by xiaocq2001** [comment]
> Since we don't do much with the `Builder` under the hood, why not just ask the user to build it first?

I just keep align with the implement for `rustls` where `ClientConfig` is passed to build connection.

File: rumqttc/src/lib.rs (hunk 1 above)
Comment (xiaocq2001) [reply rumqttc/src/lib.rs]: Close the discuss. Use TlsConnector which has debug trait.

**Comment by xiaocq2001** [comment]
Added wss (WebSocket Secure) support.
tls-native works fine connecting to mosquitto public broker.

Facing issue when using rustls, it always fail with "failed to lookup address information: Name or service not known", I guess some specific certification should be used?

**Review by de-sh — APPROVED** [review]

**Comment by xiaocq2001** [comment]
It's approved and tests are green, is there any more thing I need to do to let it merged?

BTW, is there any timeline for next rumqttc release and crate update?

**Comment by swanandx** [comment]
> It's approved and tests are green, is there any more thing I need to do to let it merged?
> 
hey, thanks for your patience and reminder :) actually we got caught up with other work and so this was missed.

> BTW, is there any timeline for next rumqttc release and crate update?
>
due to same reason of not having much bandwidth, can't commit to any specific date rn, but should be quite soon.

File: rumqttc/src/v5/eventloop.rs [hunk 3]
```diff
@@ -366,7 +366,7 @@ async fn network_connect(options: &MqttOptions) -> Result<Network, ConnectionErr
 
             Network::new(WsStream::new(socket), max_incoming_pkt_size)
         }
-        #[cfg(all(feature = "use-rustls", feature = "websocket"))]
+        #[cfg(all(any(feature = "use-rustls", feature = "use-native-tls"), feature = "websocket"))]
```
Comment (de-sh) [inline rumqttc/src/v5/eventloop.rs:369]:
```suggestion
        #[cfg(all(feature = "use-rustls", feature = "websocket"))]
```

File: rumqttc/src/v5/eventloop.rs [hunk 4]
```diff
@@ -377,6 +377,11 @@ async fn network_connect(options: &MqttOptions) -> Result<Network, ConnectionErr
                 request = request_modifier(request).await;
             }
 
+            #[cfg(feature = "use-native-tls")]
+            let connector = tls::native_tls_connector(&tls_config).await?;
+
+            // cfg: "use-native-tls" overwrites "use-rustls"
+            #[cfg(all(feature = "use-rustls", not(feature = "use-native-tls")))]
```
Comment (de-sh) [inline rumqttc/src/v5/eventloop.rs:384]:
```suggestion
```

File: rumqttc/src/eventloop.rs [hunk 5]
```diff
@@ -450,7 +450,7 @@ async fn network_connect(
                 options.max_outgoing_packet_size,
             )
         }
-        #[cfg(all(feature = "use-rustls", feature = "websocket"))]
+        #[cfg(all(any(feature = "use-rustls", feature = "use-native-tls"), feature = "websocket"))]
```
Comment (de-sh) [inline rumqttc/src/eventloop.rs:453]:
```suggestion
        #[cfg(all(feature = "use-rustls", feature = "websocket"))]
```

File: rumqttc/src/eventloop.rs [hunk 6]
```diff
@@ -461,6 +461,11 @@ async fn network_connect(
                 request = request_modifier(request).await;
             }
 
+            #[cfg(feature = "use-native-tls")]
+            let connector = tls::native_tls_connector(&tls_config).await?;
+
+            // cfg: "use-native-tls" overwrites "use-rustls"
+            #[cfg(all(feature = "use-rustls", not(feature = "use-native-tls")))]
```
Comment (de-sh) [inline rumqttc/src/eventloop.rs:468]:
```suggestion
```

File: rumqttc/examples/wss.rs
[the remark this answers is not in the document]
Comment (xiaocq2001) [reply rumqttc/examples/wss.rs:1]: Done.

**Comment by xiaocq2001** [comment]
> Please move changes regarding wss from this PR to another

OK. I'll create another PR for websocket changes after this one is merged.

**Review by swanandx — APPROVED** [review]

File: rumqttc/src/lib.rs [hunk 7]
```diff
@@ -100,7 +100,10 @@ extern crate log;
 
 use std::fmt::{self, Debug, Formatter};
 
-#[cfg(any(feature = "use-rustls", feature = "websocket"))]
+#[cfg(any(
+    feature = "use-rustls",
+    feature = "websocket"
+))]
```
Comment (de-sh) [inline rumqttc/src/lib.rs:106]:
```suggestion
#[cfg(any(feature = "use-rustls", feature = "websocket"))]
```

File: rumqttc/src/lib.rs [hunk 8]
```diff
@@ -229,8 +236,8 @@ pub enum Transport {
     #[cfg(feature = "websocket")]
     #[cfg_attr(docsrs, doc(cfg(feature = "websocket")))]
     Ws,
-    #[cfg(all(feature = "use-rustls", feature = "websocket"))]
-    #[cfg_attr(docsrs, doc(cfg(all(feature = "use-rustls", feature = "websocket"))))]
+    #[cfg(all(any(feature = "use-rustls", feature = "use-native-tls"), feature = "websocket"))]
+    #[cfg_attr(docsrs, doc(cfg(all(any(feature = "use-rustls", feature = "use-native-tls"), feature = "websocket"))))]
```
Comment (de-sh) [inline rumqttc/src/lib.rs:240]:
```suggestion
    #[cfg(all(feature = "use-rustls", feature = "websocket"))]
    #[cfg_attr(docsrs, doc(cfg(all(feature = "use-rustls", feature = "websocket"))))]
```

File: rumqttc/src/lib.rs [hunk 9]
```diff
@@ -301,8 +308,8 @@ impl Transport {
         Self::wss_with_config(config)
     }
 
-    #[cfg(all(feature = "use-rustls", feature = "websocket"))]
-    #[cfg_attr(docsrs, doc(cfg(all(feature = "use-rustls", feature = "websocket"))))]
+    #[cfg(all(any(feature = "use-rustls", feature = "use-native-tls"), feature = "websocket"))]
+    #[cfg_attr(docsrs, doc(cfg(all(any(feature = "use-rustls", feature = "use-native-tls"), feature = "websocket"))))]
```
Comment (de-sh) [inline rumqttc/src/lib.rs:312]:
```suggestion
    #[cfg(all(feature = "use-rustls", feature = "websocket"))]
    #[cfg_attr(docsrs, doc(cfg(all(feature = "use-rustls", feature = "websocket"))))]
```

**Review by de-sh — CHANGES_REQUESTED** [review]

**Review by de-sh — APPROVED** [review]

**Comment by xiaocq2001** [comment]
Thanks.
WSS updates are here:
https://github.com/bytebeamio/rumqtt/pull/881

**Pull request merged**